src/types/type_param.rs

```diff
@@ -70,49 +57,38 @@ impl TypeArg {
 /// Checks a [TypeArg] is as expected for a [TypeParam]
 pub fn check_type_arg(arg: &TypeArg, param: &TypeParam) -> Result<(), TypeArgError> {
     match (arg, param) {
-        (TypeArg::Type(_), TypeParam::Type) => Ok(()),
-        (TypeArg::ClassicType(_), TypeParam::ClassicType) => Ok(()),
-        (TypeArg::HashableType(_), TypeParam::HashableType) => Ok(()),
-        (TypeArg::List(items), TypeParam::List(ty)) => {
-            for item in items {
-                check_type_arg(item, ty.as_ref())?;
+        (
+            TypeArg::Type(SimpleType::Classic(ClassicType::Hashable(_))),
+            TypeParam::Type(TypeTag::Hashable),
+        ) => Ok(()),
+        (TypeArg::Type(SimpleType::Classic(_)), TypeParam::Type(TypeTag::Classic)) => Ok(()),
+        (TypeArg::Type(_), TypeParam::Type(TypeTag::Simple)) => Ok(()),
+        (TypeArg::List(items), TypeParam::List(param)) => {
+            items.iter().try_for_each(|arg| check_type_arg(arg, param))
+        }
+        (TypeArg::Tuple(items), TypeParam::Tuple(types)) => {
+            let items_len = items.len();
+            let types_len = types.len();
+            if items.len() != types.len() {
```

Comment:
Given you've just pulled out `items_len` and `types_len`, use them here!

fwiw I don't tend to worry much about repeated constants in error-handling code, it's exceptional so I prefer to do whatever's best for the non-exceptional case